p2p-tests/star-topology.sh

```diff
@@ -0,0 +1,143 @@
+# with Radicle Linking Exception. For full terms see the included
+# LICENSE file.
+
+set -euo pipefail
+
+function usage {
+  cat <<EOF
+Usage: sudo $(basename "${BASH_SOURCE[0]}") start | stop
+
+This script sets up a star network topology.
+It only works on Linux and _has to be run as root_.
+
+EOF
+  exit
+}
+
+function setup_network {
+  echo "setting up network: maintainer <---> seed <---> contributor"
+
+  ip netns add maintainer
+  ip netns add seed
+  ip netns add contributor
+
+  ip link add veth0 type veth peer name veth1
+  ip link add veth2 type veth peer name veth3
+  ip link add veth4 type veth peer name veth5
+
+  ip link add br0 type bridge
+
+  ip link set veth0 netns maintainer
+  ip link set veth2 netns contributor
+  ip link set veth4 netns seed
+
+  ip addr add 10.0.0.4/24 dev br0
+  ip link set dev br0 up
+
+  ip link set dev veth1 master br0
+  ip link set dev veth3 master br0
+  ip link set dev veth5 master br0
+
+  ip link set dev veth1 up
+  ip link set dev veth3 up
+  ip link set dev veth5 up
+
+  ip netns exec maintainer ip addr add 10.0.0.1/32 dev veth0
+  ip netns exec maintainer ip link set dev veth0 up
+
+  ip netns exec contributor ip addr add 10.0.0.2/32 dev veth2
+  ip netns exec contributor ip link set dev veth2 up
+
+  ip netns exec seed ip addr add 10.0.0.3/32 dev veth4
+  ip netns exec seed ip link set dev veth4 up
+
+  ip netns exec maintainer ip route flush all
+  ip netns exec contributor ip route flush all
+  ip netns exec seed ip route flush all
+
+  ip netns exec maintainer ip route add 10.0.0.3/32 dev veth0
+  ip netns exec maintainer ip route add 10.0.0.4/32 dev veth0
+  ip netns exec contributor ip route add 10.0.0.3/32 dev veth2
+  ip netns exec contributor ip route add 10.0.0.4/32 dev veth2
+  ip netns exec seed ip route add 10.0.0.1/32 dev veth4
+  ip netns exec seed ip route add 10.0.0.2/32 dev veth4
+  ip netns exec seed ip route add 10.0.0.4/32 dev veth4
+}
+
+function clean_up_network {
+  echo "tearing down network"
+  set +e
+
+  # Force commands to execute even if there are errors.
+  ip netns delete maintainer > /dev/null 2>&1
+  ip netns delete seed > /dev/null 2>&1
+  ip netns delete contributor > /dev/null 2>&1
+
+  ip link delete veth0 > /dev/null 2>&1
+  ip link delete veth1 > /dev/null 2>&1
+  ip link delete veth2 > /dev/null 2>&1
+  ip link delete veth3 > /dev/null 2>&1
+  ip link delete veth4 > /dev/null 2>&1
+  ip link delete veth5 > /dev/null 2>&1
+
+  ip link delete br0 > /dev/null 2>&1
+
+  # TODO: figure out a less invasive way to make this work on CI!
+  iptables -P INPUT ACCEPT
+  iptables -P FORWARD ACCEPT
+  iptables -P OUTPUT ACCEPT
+  iptables -t filter -F
+  iptables -t filter -X
+  iptables -t nat -F
+  iptables -t nat -X
+  iptables -t mangle -F
+  iptables -t mangle -X
+  iptables -t raw -F
+  iptables -t raw -X
+
+  set -e
+  return 0
+}
+
+echo
+
+if [ "$EUID" -ne 0 ]
+  then usage;
+  exit 1
+fi
+
+function status {
+  echo "current networking status"
```

Comment:
Can we use `set +x` here too?

Reply:
61ad508c6db101dfe536c64e429190b1bbcfa869